### PR TITLE
Rewrite of VicarLabel and VicarImage, improve docs

### PR DESCRIPTION
**Pull request by markshowalter**
This pull request contains significant rewrites of VicarLabel and VicarImage to address a few bugs, simplify, and remove a performance logjam that I knew about (in which adding one new parameter at a time to a label took N-squared time, where N is the number of parameters in the label).

# Fixed Issues

- Failure to read a Galileo SSI file containing a non-ASCII string.

# Summary of Changes

## New handling of non-compliant VICAR labels

We had already programmed around a VICAR compliance issue in Cassini ISS files, where one VICAR parameter name is 33 characters long, whereas the documentation says the limit is 32. Now we see that there is at least one Galileo SSI file with a different violation of the standard. This makes me uncomfortable. To make an analogy, nobody would ask the developer of the Python FITS module to change their code so it could read a file that violates the FITS standard.

My new solution is to add a "`strict`" parameter to the constructors, with a default value of `True`. If you are dealing with a non-compliant file, you will need to specify "`strict=False`". Then it uses a looser interpretation of the standard. In the future, if we encounter more non-compliant files, we can look at the nature of those and perhaps make them also compatible with our loose parsing rules. But meanwhile, strict parsing is always the default. In the case of Cassini ISS, the one change needed to the host module is to specify "`strict=False`" in the `VicarImage` constructor.

I added the Galileo SSI file with a corrupt string value in the label to `test_files/`.

## Performance logjam

Previously, every time you modified a VicarLabel, it regenerated an internal index. Because there is a strong tendency to add parameters to a label one at a time, this causes an "N-squared" performance issue. I added an internal flag so that the label no longer re-indexes until it needs to.

## `make html` warning

You get an annoying warning on "`make html`" inside the `docs` about a missing `_static` subdirectory. I created an empty `docs/_static` subdirectory and voila, the message goes away.

## Approach to unit tests

These are now consistent with the rest of Python-world.

- `tests/__init__.py` now has no content.
- No test files contain the "`if __name__ == '__main__`" clause at the bottom.

## Revised docstrings

My high-level goal is to make sure the same content is available to a user in a Python session and to a readthedocs user. I give the Sphinx/Napoleon developers credit for how easy it is to make clean HTML documentation from a docstring that also remains perfectly readable. Frankly, however, I would expect to continue using docstrings and `help()` more often than readthedocs webpages; I only tend to go to HTML documentation when I'm a new user or when `dir()` and `help()` don't get me to what I need. So I want to make sure going forward that both sources of documentation are complete. To that end...

- Most of the content in `README.md` is now duplicated in the docstring for `vicar/__init__.py`. This is needed so the content shows up in `help(vicar)`. Rob points out that keeping the two files (same content but different formats) in sync is going to be hard, and I agree. However, the `README` file is not immediately accessible to someone working in a Python shell, whereas `help()` is. If the two files drift out of sync, I understand, but I think the goal should be equivalence.

- I started using "Parameters:" instead of "Args:" in docstrings because it better matches the HTML that Sphinx generates.

- Sphinx does not cleanly handle the case of a function returning multiple values in a tuple, or (even worse) a function that returns a variable number of results. Expanding on what Rob has been doing, I found what works best is to use bullets to explain the individual returns. See the docstrings and readthedocs for examples.

- There's a fundamental difference between the "Note:" and "Notes:" directives. The former creates a special box in the HTML to highlight something important, whereas the latter is just a header for a new section of text, nothing more. Keep this difference in mind. If it's an important point that might be overlooked by the user, use "Note:". If it's just some more explanatory text, use "Notes:".

- I added the `:special-members:` directive to `docs/module.rst`, because this includes the dunder functions like `__len__`, `__getitem__`,` __eq__`, etc. in the HTML. These are an important part of the public API and are otherwise not included. Unfortunately, this also has the side-effect of including a few things we don't want like `__dict__` and `__hash__`, so I have added these in an `:exclude-members:` directive. That seems to produce all the public API documentation we want for readthedocs.

- Meanwhile, I think it will be useful also to have a developer's version of the HTML documentation. This again echos my own use case, where I look at docstrings for private functions (those beginning with an underscore) every bit as often as for public functions. I created a new directory `docs-dev`, which is a copy of `docs/` but `module.rst` is modified to include private functions (using `:private-members:`) and also to include `_LABEL_GRAMMAR.py`, which is a core piece of the module. This extra content is not appropriate for readthedocs, but it would be easy and useful for us to host this web content locally. Details of that are TDB. Another advantage of building this version of the HTML is that it's the only way to find out if the docstrings of the private functions are properly formatted.

- Minor fix: JPL's MIPL website has disappeared, mysteriously, and with it the document that describes the VICAR standard. Luckily, we host a local copy in our website's `help/` subdirectory, so I changed the links in the documentation to point to our copy.

# Known Problems

- We're going to have to add that one line fix to oops/hosts/cassini/iss for `strict=False`.

### CODE REVIEW

**Comment by markshowalter** [comment]
More changes... I moved the definitions at the top of `vicarlabel.py` and `vicarimage.py` into a separate file `_DEFINITIONS.py`, in part so many of the `flake8` violations were more localized. The docstrings in `_DEFINITIONS.py` and `_LABEL_GRAMMAR.py` now contain links to the source on GitHub, which makes up for the fact that the `viewsource` module doesn't support files just filled with definitions. Finally, I stripped `.flake8` back to its bare minimum.

File: README.md [hunk 1]
```diff
@@ -25,7 +25,7 @@
 # Introduction
 
 `vicar` is a Python module that supports reading and writing of JPL's VICAR file format. It supports the definition of the VICAR file format as found here:
-[https://www-mipl.jpl.nasa.gov/external/VICAR_file_fmt.pdf](https://www-mipl.jpl.nasa.gov/external/VICAR_file_fmt.pdf)
+[https://pds-rings.seti.org/help](https://pds-rings.seti.org/help)
```
Comment (rfrenchseti) [inline README.md]: This needs to be the full filename: https://pds-rings.seti.org/help/VICAR_file_fmt.pdf

File: tests/test_vicarimage.py [hunk 2]
```diff
@@ -141,18 +143,18 @@ def test_VicarImage(self):
 
         # binheader
         test = VicarImage(test_dir / 'C2069302_GEOMA.DAT')
-        self.assertIsNone(test.array)
-        self.assertIsNone(test.array2d)
-        self.assertIsNone(test.array3d)
-        self.assertIsNone(test.prefix)
-        self.assertIsNone(test.prefix2d)
-        self.assertIsNone(test.prefix3d)
+        self.assertTrue(test.array is None)
```
Comment (rfrenchseti) [inline tests/test_vicarimage.py]: Why do you not like `self.assertIsNone`?

File: vicar/_LABEL_GRAMMAR.py [hunk 3]
```diff
@@ -2,69 +2,84 @@
 # vicar/_LABEL_GRAMMAR.py
 ##########################################################################################
 """\
+Definition of _LABEL_GRAMMAR, a `pyparsing` grammar supporting the VICAR format standard,
+used internally. The full standard is defined here:
+
+    https://www-mipl.jpl.nasa.gov/external/VICAR_file_fmt.pdf
```
Comment (rfrenchseti) [inline vicar/_LABEL_GRAMMAR.py]: This is still the bad URL

File: docs-dev/module.rst [hunk 4]
```diff
@@ -0,0 +1,22 @@
+``vicar`` Module
+================
+
+.. automodule:: vicar
+    :member-order: groupwise
+    :members:
+    :undoc-members:
+    :special-members:
+    :private-members:
+    :exclude-members: __dict__, __hash__, __module__, __weakref__
```
Comment (rfrenchseti) [inline docs-dev/module.rst:10]: These excludes need to be added to the main `docs/module.rst` as well. Right now things like `__dict__` and `__weakref__` are showing up in the main documentation.

File: vicar/vicarimage.py [hunk 5]
```diff
@@ -3,183 +3,144 @@
...
-    """This class defines the contents of a VICAR data file. It supports methods for
-    reading and writing files and for accessing header information.
+    """Constructor for a VicarImage.
+
+    This class defines the contents of a VICAR data file. It supports methods for reading
+    and writing files and for accessing header information.
+
+    **Properties**:
+
+        * array: The 3-D data array converted to native format.
```
Comment (rfrenchseti) [inline vicar/vicarimage.py]: Do you want to put all these method names inside back-ticks?

File: vicar/vicarimage.py [hunk 6]
```diff
@@ -714,385 +833,565 @@ def _read_file(filepath, extraneous='ignore'):
...
+                4                 = ("", 0, 0, 4)     -> "TEXP=1.5    "
+                (1, 4)            = ("", 0, 1, 4)     -> "TEXP= 1.5    "
+                (2, 1, 4)         = ("", 2, 1, 4)     -> "TEXP  = 1.5    "
 
-        After the value, the optional items are:
+            When the parameter value is a list, it is also possible to embed formatting
+            information on an item by item basis. Replace any item value by a tuple:
 
-            format: a format string, e.g., "%+07d", "%12.3e" or "%.4f".
+                (`item`[, `format`][[, `blanks_before`], `blanks_after`])
```
Comment (rfrenchseti) [inline vicar/vicarimage.py]: Some of these back-ticks are showing up in the HTML

File: vicar/vicarimage.py [hunk 7]
```diff
@@ -198,38 +159,54 @@ def __init__(self, source=None, array=None, *, prefix=None, binheader=None):
...
-            VicarLabel: The label associated with this VicarImage."""
+            VicarLabel: The VicarLabel object.
+        """
+
         return self._label
 
     @property
     def array2d(self):
-        """Return the array object as 2-D.
+        """The data array as 2-D.
```
Comment (rfrenchseti) [inline vicar/vicarimage.py:203]: I don't know why, but none of the properties are showing up with a [Source] link

File: vicar/vicarimage.py [hunk 8]
```diff
@@ -452,36 +472,106 @@ def binheader(self, value):
...
+                True (the default) to require strict conformance to the VICAR standard;
+                False for a looser version of the standard.
 
         Returns:
-            VicarImage: The created VicarImage object.
+            VicarImage or (VicarImage, bytes or None): A VicarImage or a tuple containing:
+
+            * VicarImage: A new object containing the content of the specified file.
+            * bytes or None: Any extraneous bytes from the end of the file, included if
+              if `extraneous` equals "include".
```
Comment (rfrenchseti) [inline vicar/vicarimage.py]: "if if"

**Review by rfrenchseti — CHANGES_REQUESTED** [review]

File: vicar/vicarimage.py (hunk 7 above)
Comment (markshowalter) [reply vicar/vicarimage.py:203]: This appears to be a known bug. Here's the issue in the GitHub repo for sphinx.ext.viewcode.
[https://github.com/sphinx-doc/sphinx/issues/11279](https://github.com/sphinx-doc/sphinx/issues/11279)
This is not worth worrying about till they fix it.

File: vicar/vicarimage.py (hunk 6 above)
Comment (markshowalter) [reply vicar/vicarimage.py]: You can't put the second backtick just before a left bracket, apparently. I needed to insert a space before the bracket.

File: vicar/vicarimage.py (hunk 5 above)
Comment (markshowalter) [reply vicar/vicarimage.py]: Done

File: docs-dev/module.rst (hunk 4 above)
Comment (markshowalter) [reply docs-dev/module.rst:10]: Done

File: vicar/_LABEL_GRAMMAR.py (hunk 3 above)
Comment (markshowalter) [reply vicar/_LABEL_GRAMMAR.py]: Fixed

File: tests/test_vicarimage.py (hunk 2 above)
Comment (markshowalter) [reply tests/test_vicarimage.py]: Changed.

File: README.md (hunk 1 above)
Comment (markshowalter) [reply README.md]: Fixed

**Review by rfrenchseti — APPROVED** [review]

**Pull request merged**